### PR TITLE
Remove qhttp; replace with curl and qhttpengine

### LINKED ISSUE #913: Remove qhttp and KeePassHTTP functionality
When I create any entry for a website for a known favicon, clicking "download favicon" reports the error " unable to fetch favicon" in all cases
## Expected Behavior

I expect clicking "download favicon" to correctly pull the icon from the database
## Current Behavior

ERROR: unable to fetch favicon
## Steps to Reproduce (for bugs)

1. Create or edit an existing entry for a website with a known favicon
2. Click icon on the left side
3. Click download favicon on the bottom

## Context

I would have to manually import every favicon with a file instead of having them being pulled from the database
## Debug Info

KeePassXC - 2.20
Revision:  caa49a8ef3ee28ed478192389b21d61107b3b8e0

Libraries:
- Qt 5.9.0
- libgcrypt 1.7.7

Operating system: Windows 10 
CPU architecture: 64 bit
Kernel: Windows NT

Enabled extensions:
- KeePassHTTP
- Auto-Type
- YubiKey
[2017-08-30 14-50-17.zip](https://github.com/keepassxreboot/keepassxc/files/1264715/2017-08-30.14-50-17.zip)

### PR DESCRIPTION
**Pull request by droidmonkey**
## Description

This PR completely removes qhttp and replaces with the following packages:

* Use curl for http client requests (Custom Icons)
* Use [qhttpengine ](https://github.com/nitroshare/qhttpengine) for KeePassHTTP server

The custom icon download sequence was significantly improved with code reduction around 60% and no recursion.

*Note: This PR requires the debian repositories to be updated with qhttpengine 1.x before merging*

Fixes #913 

## Motivation and context

The primary driver for this change is to align with Debian requirements. Secondarily, we are moving off of an unsupported, flaky library.

## How has this been tested?

Manually; automated test cases should be written for this when KeePass-Browser is merged.

## Types of changes

- ✅ Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- ✅ I have read the **CONTRIBUTING** document. **[REQUIRED]**
- ✅ My code follows the code style of this project. **[REQUIRED]**
- ✅ All new and existing tests passed. **[REQUIRED]**
- ✅ I have compiled and verified my code with `-DWITH_ASAN=ON`. **[REQUIRED]**
- ✅ My change requires a change to the documentation and I have updated it accordingly.
- ❌  I have added tests to cover my changes.

### CODE REVIEW

**Comment by droidmonkey** [comment]
Please put the review on hold while I figure out these dependency issues...

**Comment by nathan-osman** [comment]
If I may jump in and offer a bit of help... QHttpEngine wasn't added to Ubuntu until Yakkety (16.10) was released. Unfortunately, Travis CI is still using Trusty (14.04) for their default build environment.

I do maintain a PPA with builds for Xenial (16.04) but apparently Travis CI doesn't provide an option for building on Xenial yet (even though it was released over a year and a half ago).

Apparently it may be possible to use Xenial in Docker on Travis CI. I'm not entirely sure how that would work but it seems to be the only option.

**Comment by nathan-osman** [comment]
After giving this some more thought - one other option is to stick with Trusty, add [this PPA](https://launchpad.net/~beineri/+archive/ubuntu/opt-qt542-trusty), and then download/build/install QHttpEngine directly.

I use a similar approach for testing QHttpEngine itself in Travis CI - you might find this useful: https://github.com/nitroshare/qhttpengine/blob/master/.travis.yml

**Comment by droidmonkey** [comment]
Hi @nathan-osman. Unfortunately the distribution hoops and significant difference in version (0.1 vs 1.0) are leading me to have to go back to libmicrohttpd. I didn't realize there was such a difference between version ticks. Would it be possible to smooth out what is already in the downstream repos, especially Debian?

**Comment by nathan-osman** [comment]
I'd love to but I don't have package upload rights to Debian so I'm dependent on getting a sponsor to upload new versions of the package. And yes, there were some rather significant changes between versions which makes this more complicated since the new version will probably a new package name to avoid breaking existing packages.

One option is always just embedding QHttpEngine in your project's source code. It is released under the MIT license so it should be usable pretty much anywhere.

**Comment by droidmonkey** [comment]
One of the main drivers in this move to your engine is Debian requiring us to move the http libraries outside the code base (ie, in a dynamically linked library) to allow for security updates and comply with their policies. We have a line in with the Debian folks, it might be easy to just append a modifier like ```qhttpengine-1-dev```

**Comment by adolfogc** [comment]
@droidmonkey I tried using the following lines in `src/http/CMakeLists.txt`:

```cmake
# ...
    if(CMAKE_SYSTEM_NAME STREQUAL "Linux")
      find_package(QHttpEngine REQUIRED CONFIG)
      set(qhttpengine QHttpEngine)
      find_package(CURL REQUIRED)
      set(curl CURL)
    else()
      find_package(qhttpengine REQUIRED CONFIG)
      find_package(curl REQUIRED)
    endif()
# ...
```

It was able to find `QHttpEngine` in MSYS2/macOS/Linux (Docker image based on debian:stretch); it wasn't able to find `curl` though.

File: src/gui/EditWidgetIcons.cpp [hunk 1]
```diff
@@ -159,107 +153,79 @@ void EditWidgetIcons::setUrl(const QString& url)
...
+    }
 
-    m_ui->faviconButton->setDisabled(true);
+    m_ui->faviconButton->setDisabled(false);
+#endif
 }
 
-void EditWidgetIcons::fetchFaviconFromGoogle(const QString& domain)
+#ifdef WITH_XC_HTTP
+size_t writeCurlResponse(char *ptr, size_t size, size_t nmemb, void *data)
```
Comment (phoerious) [inline src/gui/EditWidgetIcons.cpp]: Pointer operator belongs to type.

File: src/gui/EditWidgetIcons.cpp [hunk 2]
```diff
@@ -159,107 +153,79 @@ void EditWidgetIcons::setUrl(const QString& url)
...
-        resetFaviconDownload();
-        m_fallbackToGoogle = false;
-        QUrl faviconUrl = QUrl("https://www.google.com/s2/favicons");
-        faviconUrl.setQuery("domain=" + QUrl::toPercentEncoding(domain));
-        fetchFavicon(faviconUrl);
-    } else {
-        resetFaviconDownload();
-        emit messageEditEntry(tr("Unable to fetch favicon."), MessageWidget::Error);
-    }
+    QByteArray* res = static_cast<QByteArray*>(data);
```
Comment (phoerious) [inline src/gui/EditWidgetIcons.cpp]: No need to shorten variable names like that here.

**Comment by droidmonkey** [comment]
@adolfogc careful, the qhttpengine located in the linux distros is the old version and incompatible with the v1.0 that I coded against. Working with the package maintainer to fix that issue.

**Comment by adolfogc** [comment]
@droidmonkey 

> Working with the package maintainer to fix that issue.

Great!

**Comment by ArchangeGabriel** [comment]
What the status of this PR? Is it still relevant with the HTTP module being dropped in the (quite) near future? Maybe the curl part is still important?

**Comment by phoerious** [comment]
Yes, this is still scheduled for 2.3. qhttp is just too buggy.

**Comment by yan12125** [comment]
Hello, great to see there's an update. I have a question and feedback:

About qhttpengine:

> Unfortunately the distribution hoops and significant difference in version (0.1 vs 1.0) are leading me to have to go back to libmicrohttpd

Is there still a plan to go back to libmicrohttpd? I'm not sure if I should create a qhttpengine package for MacPorts (I'm the maintainer of KeePassXC at MacPorts).

About curl:

Just like the case in https://github.com/keepassxreboot/keepassxc/pull/1169#issuecomment-345827600, on Arch Llinux I need to change ```find_package(curl REQUIRED)``` to ```find_package(CURL REQUIRED)``` so that the curl package can be found. I guess it's because the relevant CMake module is called FindCURL.cmake instead of Findcurl.cmake [1].

[1] https://gitlab.kitware.com/cmake/cmake/blob/master/Modules/FindCURL.cmake

**Comment by droidmonkey** [comment]
libmicrohttpd is off the table, I want to stick with the Qt solution.

**Comment by yan12125** [comment]
Hello, could you rebase this? The yubikey fix is merged into develop (d9119877872101493dbf5ad44d14afcd10afcd26) and leads to a conflict.

**Comment by droidmonkey** [comment]
closing in favor of a new PR that only implements CURL.

**Pull request closed without merging**